src/java.base/share/classes/jdk/internal/classfile/impl/SplitConstantPool.java

```diff
@@ -0,0 +1,604 @@
+import static jdk.internal.classfile.Classfile.TAG_MODULE;
+import static jdk.internal.classfile.Classfile.TAG_NAMEANDTYPE;
+import static jdk.internal.classfile.Classfile.TAG_PACKAGE;
+import static jdk.internal.classfile.Classfile.TAG_STRING;
+
+/**
+ * ConstantPool.
+ */
+public final class SplitConstantPool implements ConstantPoolBuilder {
+
+    private final ClassReaderImpl parent;
+    private final int parentSize, parentBsmSize;
+    final Options options;
+
+    private int size, bsmSize;
+    private PoolEntry[] myEntries;
+    private BootstrapMethodEntryImpl[] myBsmEntries;
+    private boolean doneFullScan;
+    private EntryMap<PoolEntry> map;
+    private EntryMap<BootstrapMethodEntryImpl> bsmMap;
+
+    public SplitConstantPool() {
+        this(new Options(Collections.emptyList()));
+    }
+
+    public SplitConstantPool(Options options) {
+        this.size = 1;
+        this.bsmSize = 0;
+        this.myEntries = new PoolEntry[1024];
+        this.myBsmEntries = new BootstrapMethodEntryImpl[8];
+        this.parent = null;
+        this.parentSize = 0;
+        this.parentBsmSize = 0;
+        this.options = options;
+    }
+
+    public SplitConstantPool(ClassReader parent) {
+        this.options = ((ClassReaderImpl) parent).options;
+        this.parent = (ClassReaderImpl) parent;
+        this.parentSize = parent.entryCount();
+        this.parentBsmSize = parent.bootstrapMethodCount();
+        this.size = parentSize;
+        this.bsmSize = parentBsmSize;
+        this.myEntries = new PoolEntry[8];
+        this.myBsmEntries = new BootstrapMethodEntryImpl[8];
+    }
+
+    @Override
+    public int entryCount() {
+        return size;
+    }
+
+    @Override
+    public int bootstrapMethodCount() {
+        return bsmSize;
+    }
+
+    @Override
+    public PoolEntry entryByIndex(int index) {
+        return (index < parentSize)
+               ? parent.entryByIndex(index)
+               : myEntries[index - parentSize];
+    }
+
+    @Override
+    public BootstrapMethodEntryImpl bootstrapMethodEntry(int index) {
+        return (index < parentBsmSize)
+               ? parent.bootstrapMethodEntry(index)
+               : myBsmEntries[index - parentBsmSize];
+    }
+
+    public Options options() {
+        return options;
+    }
+
+    @Override
+    public boolean canWriteDirect(ConstantPool other) {
+        return this == other || parent == other;
+    }
+
+    @Override
+    public boolean writeBootstrapMethods(BufWriter buf) {
+        if (bsmSize == 0)
+            return false;
+        int pos = buf.size();
+        if (parent != null && parentBsmSize != 0) {
+            parent.writeBootstrapMethods(buf);
+            for (int i = parentBsmSize; i < bsmSize; i++)
+                bootstrapMethodEntry(i).writeTo(buf);
+            int attrLen = buf.size() - pos;
+            buf.patchInt(pos + 2, 4, attrLen - 6);
+            buf.patchInt(pos + 6, 2, bsmSize);
+            return true;
```

Comment:
The if and else branch return true, factor out at the end of the method?

Reply:
fixed, thanks.